optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsBiCondition.java

```diff
@@ -62,31 +74,21 @@ public DroolsBiCondition(DroolsMetadata<Object, A> aMetadata, DroolsMetadata<Obj
     }
 
     public DroolsBiCondition<A, B> andFilter(BiPredicate<A, B> predicate) {
-        PatternDSL.PatternDef<Object> newPattern = bMetadata.getPattern()
+        Supplier<PatternDSL.PatternDef<Object>> patternSupplier = () -> bMetadata.buildPattern()
```

Comment:
Does the `.buildPattern()` and `aMetadata.getVariableDeclaration()` calls need to be called lazely?
I don't grok the entire code, but can you verify that this is doesn't work:

```
bBuildPattern = bMetadata.buildPattern();
Supplier<PatternDSL.PatternDef<Object>> patternSupplier = () -> bBuildPattern.expr(...);
```

If it does, it's more efficient. Some question for `aMetadata.getVariableDeclaration()`. Clearly, the predicate test must be called later and cannot be extracted from the supplier, because it uses a and b.

Reply:
The reason for all this laziness, which I don't like, is so that we can reuse patterns in two different places. Use case is in the new rule in this PR:

- I want to use the pattern in the first accumulate.
- And then I want to use the same pattern, with some modifications (a new filter) in the second accumulate.
- Drools does not let me to clone patterns.
- Therefore, I need to create a new pattern every time I use it.